File: packages/main-api/src/types/space_visibility.rs
use bdk::prelude::*;

use super::Visibility;

#[derive(
    Debug,
    Clone,
    serde_with::SerializeDisplay,
    serde_with::DeserializeFromStr,
    Default,
    DynamoEnum,
    JsonSchema,
    PartialEq,
)]
pub enum SpaceVisibility {
    #[default]
    Private,
    Public,
    // Only team members can access
    Team(String),

    // Only members in the specific team group can access
    TeamGroupMember(String),
}

impl Into<Visibility> for SpaceVisibility {
    fn into(self) -> Visibility {
        match self {
            SpaceVisibility::Private => Visibility::Private, // Private spaces are treated as Public in post visibility context
            SpaceVisibility::Public => Visibility::Public,
            SpaceVisibility::Team(team_pk) => Visibility::TeamOnly(team_pk),
            SpaceVisibility::TeamGroupMember(team_pk) => Visibility::TeamOnly(team_pk),
        }
    }
}

File: flutter-packages/mobile/lib/services/space/types/space_visibility.dart
enum SpaceVisibility { private, public, team, teamGroupMember }

SpaceVisibility spaceVisibilityFromJson(dynamic v) {
  if (v == null) return SpaceVisibility.private;

  final s = v.toString().toLowerCase();

  if (s == 'public') return SpaceVisibility.public;
  if (s.startsWith('team_group_member')) {
    return SpaceVisibility.teamGroupMember;
  }
  if (s.startsWith('team')) {
    return SpaceVisibility.team;
  }
  return SpaceVisibility.private;
}

String spaceVisibilityToJson(SpaceVisibility v) {
  switch (v) {
    case SpaceVisibility.private:
      return 'private';
    case SpaceVisibility.public:
      return 'public';
    case SpaceVisibility.team:
      return 'team';
    case SpaceVisibility.teamGroupMember:
      return 'team_group_member';
  }
}
